core/shared/src/main/scala/org/http4s/Message.scala

```diff
@@ -204,6 +207,55 @@ sealed trait Message[+F[_]] extends Media[F] { self =>
   def isChunked: Boolean =
     headers.get[`Transfer-Encoding`].exists(_.values.contains_(TransferCoding.chunked))
 
+  /** Puts a `Content-Length` header, replacing any existing.  Removes
+    * any existing `chunked` value from the `Transfer-Encoding`
+    * header.  It is critical that the supplied content length
+    * accurately describe the length of the body stream.
+    *
+    * {{{
+    * scala> import org.http4s.headers._
+    * scala> val chunked = Request().withHeaders(
+    *      |   `Transfer-Encoding`(TransferCoding.chunked),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> chunked.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res0: Headers = Headers(Content-Type: text/plain, Content-Length: 1024)
+    *
+    * scala> val chunkedGzipped = Request().withHeaders(
+    *      |   `Transfer-Encoding`(TransferCoding.chunked, TransferCoding.gzip),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> chunkedGzipped.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res1: Headers = Headers(Transfer-Encoding: gzip, Content-Type: text/plain, Content-Length: 1024)
+    *
+    * scala> val const = Request().withHeaders(
+    *      |   `Content-Length`(2048),
+    *      |   `Content-Type`(MediaType.text.plain))
+    * scala> const.withContentLength(`Content-Length`.unsafeFromLong(1024)).headers
+    * res1: Headers = Headers(Content-Type: text/plain, Content-Length: 1024)
+    * }}}
+    */
+  def withContentLength(contentLength: `Content-Length`): SelfF[F] =
```

Comment:
I don't think this fits well with the new entity model. It's the job of `Entity` to know about content length, not `Message` - anything in `Message` relating to content length should just be a thin wrapper around `Entity` - and it shouldn't be possible to set an incorrect length on a strict entity, so trying to do so should fail either at compile time or run time.

At the very least, I think this method should have `unsafe` or `unchecked` in its name.

Assuming the change in 0.23 hasn't been released yet, could we consider adjusting this in 0.23 to be more forward-compatible with 1.0's model?

Sorry to be late to the party on this 😬

Reply:
Good point, that makes sense. Pretty sure this hasn't been released yet, I'm curious what you have in mind :)

Reply:
I don't have an immediate good idea. Does this need to be public, or could it be made package-private and released as-is, so we can think later on about what a public version should look like?

Reply:
> At the very least, I think this method should have unsafe or unchecked in its name.

I opine that prefixes 'unsafe' don't work. But yeah, it'd be cool if Scala has a `@unsafe` annotation to mark unsafe code and special compiler flags to configure behavior - compile unsafe code, emit a warning, or a compilation error.

Reply:
That could definitely be done as a scalafix or wartremover rule!

Reply:
I think in `main` what we should have is `def transformEntity(f: Entity[F] => Entity[F]): SelfF[F]` (might need an extra type parameter to compile) - then the user can change the size where it makes sense to do so.

Reply:
I drafted a PR with a slightly different idea - https://github.com/http4s/http4s/pull/6285